crates/gitbutler-branch-actions/src/branch.rs

```diff
@@ -0,0 +1,408 @@
+        if let Some(group) = groups.get_mut(&identity) {
+            group.push(branch);
+        } else {
+            groups.insert(identity, vec![branch]);
+        }
+    }
+    let config = repo.config()?;
+    let local_author = Author {
+        name: config.get_string("user.name").ok(),
+        email: config.get_string("user.email").ok(),
+    };
+
+    // Convert to Branch entries for the API response, filtering out any errors
+    let branches: Vec<BranchListing> = groups
+        .iter()
+        .filter_map(|(identity, group_branches)| {
+            let branch_entry = branch_group_to_branch(
+                identity.clone(),
+                group_branches.clone(),
+                repo,
+                &local_author,
+            );
+            if branch_entry.is_err() {
+                tracing::warn!(
+                    "Failed to process branch group {:?} to branch entry: {:?}",
+                    identity,
+                    branch_entry
+                );
+            }
+            branch_entry.ok()
+        })
+        .collect();
+    Ok(branches)
+}
+
+/// Converts a group of branches with the same identity into a single branch entry
+fn branch_group_to_branch(
+    identity: Option<BString>,
+    group_branches: Vec<&GroupBranch>,
+    repo: &git2::Repository,
+    local_author: &Author,
+) -> Result<BranchListing> {
+    let virtual_branch = group_branches
+        .iter()
+        .filter_map(|branch| match branch {
+            GroupBranch::Virtual(vb) => Some(vb),
+            _ => None,
+        })
+        .next();
+    let remote_branches: Vec<&gix::Reference> = group_branches
+        .iter()
+        .filter_map(|branch| match branch {
+            GroupBranch::Remote(gb) => Some(gb),
+            _ => None,
+        })
+        .filter(|reference| matches!(reference.target(), gix::refs::TargetRef::Peeled(_)))
+        .collect();
+    let local_branches: Vec<&gix::Reference> = group_branches
+        .iter()
+        .filter_map(|branch| match branch {
+            GroupBranch::Local(gb) => Some(gb),
+            _ => None,
+        })
+        .filter(|reference| matches!(reference.target(), gix::refs::TargetRef::Peeled(_)))
+        .collect();
+
+    // Virtual branch associated with this branch
+    let virtual_branch_reference = virtual_branch.map(|branch| VirtualBranchReference {
+        given_name: branch.name.clone(),
+        id: branch.id,
+        in_workspace: branch.in_workspace,
+    });
+
+    let mut remotes: Vec<BString> = Vec::new();
+    for reference in remote_branches.iter() {
+        let short_name: String = reference.name().shorten().to_str_lossy().to_string();
+        let file_name: String = reference.name().file_name().to_str_lossy().to_string();
+        // remove file_name suffix from short_name and also remote the trailing '/' character
+        let remote_name = short_name
+            .trim_end_matches(&file_name)
+            .trim_end_matches('/');
```

Comment:
Does this account for remote names which contain forward slashes?

Reply:
@Caleb-T-Owens it turns out that "file_name" is also not what i thought it was... this will not work

Reply:
Definitely not.

I don't think that there is anything built-in for the kind of name you are looking for, i.e. `foo/bar` in a ref names `refs/remotes/my/special/remote/foo/bar` where `my/special/remote` is a valid remote name.